### PR TITLE
XLA runtime support for fused attention

### PR DESCRIPTION
**Pull request by AyanmoI**
This PR add XLA Runtime support for cuDNN runtime based Fused attention feature in XLA.

### CODE REVIEW

File: xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc [hunk 1]
```diff
@@ -526,6 +526,338 @@ class CholeskyOpLowering : public OpRewritePattern<CholeskyOp> {
...
+    call->setAttr(b.getStringAttr("uid"), b.getI64IntegerAttr(uid_.uid()));
+
+    // Helper functins to copy attributes from the conv op to the custom call.
+    auto set_attr = [&](StringRef name, Attribute attr) {
+      call->setAttr(b.getStringAttr(name), attr);
+    };
+
+    set_attr("fmha_scale", op.getFmhaScaleAttr());
+    if (op.getDropoutRate()) {
+      set_attr("dropout_rate", op.getDropoutRateAttr());
```
Comment (ezhulenev) [inline xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc:670]: You can add `if (attr) ...` check to `set_attr` and remove some of the branches here, should be more dense code and easier to read

File: xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc [hunk 2]
```diff
@@ -526,6 +526,338 @@ class CholeskyOpLowering : public OpRewritePattern<CholeskyOp> {
...
+              "BMM_Bias_Softmax_Dropout_BMM");
+        break;
+
+      case mlir::lmhlo_gpu::FusedMhaBackwardDagSignature::
+          BackwardScaleBiasMaskSoftmax:
+        if (num_operands == 11) {
+          fused_attention += "scale.mask.softmax";
+        } else if (num_operands == 12) {
+          fused_attention += "scale.dbias.mask.softmax";
+        } else
```
Comment (ezhulenev) [inline xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc:781]: add braces to all `else` cases in this file, it should be consistent between all branches in the if statement

File: xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc [hunk 3]
```diff
@@ -526,6 +526,338 @@ class CholeskyOpLowering : public OpRewritePattern<CholeskyOp> {
...
+    // Assign a unique id to this instance of a fused_attention operation.
+    call->setAttr(b.getStringAttr("uid"), b.getI64IntegerAttr(uid_.uid()));
+
+    // Helper functins to copy attributes from the conv op to the custom call.
+    auto set_attr = [&](StringRef name, Attribute attr) {
+      call->setAttr(b.getStringAttr(name), attr);
+    };
+
+    set_attr("fmha_scale", op.getFmhaScaleAttr());
+    if (op.getDropoutRate()) {
```
Comment (ezhulenev) [inline xla/mlir/backends/gpu/transforms/lmhlo_gpu_to_gpu_runtime.cc:820]: and the same for `set_attr` as above

File: xla/translate/mhlo_to_lhlo_with_xla/mhlo_to_lhlo_with_xla.cc [hunk 4]
```diff
@@ -1527,8 +1527,8 @@ tsl::StatusOr<Operation*> LhloDialectEmitter::EmitDnnfMHABackward(
         builder_, config.bmm2_grad_gemm2_dot_dimension_numbers()));
 
     op.setFmhaScaleAttr(builder_.getF64FloatAttr(config.fmha_scale()));
-    op.setDropoutRateAttr(builder_.getF64FloatAttr(config.dropout_rate()));
-    op.setSeedAttr(builder_.getI64IntegerAttr(config.seed()));
+    // op.setDropoutRateAttr(builder_.getF64FloatAttr(config.dropout_rate()));
```
Comment (ezhulenev) [inline xla/translate/mhlo_to_lhlo_with_xla/mhlo_to_lhlo_with_xla.cc:1530]: Leftover? Should be removed or uncmmmended.

**Comment by AyanmoI** [comment]
@ezhulenev Looks clean?

**Review by ezhulenev — APPROVED** [review]

**Comment by dathudeptrai** [comment]
@AyanmoI Just wonder if this is something like FlashAttention ?

**Comment by AyanmoI** [comment]
> @AyanmoI Just wonder if this is something like FlashAttention ?

Flash Attention work is in progress.

**Comment by dathudeptrai** [comment]
@AyanmoI does it mean in the near future of TF release, flash-attention can be applied with @tf.function(jit_compile=True) ?

**Comment by AyanmoI** [comment]
> @AyanmoI does it mean in the near future of TF release, flash-attention can be applied with @tf.function(jit_compile=True) ?

That is correct. XLA is expected to be able to lower to cuDNN flash attention kernel in the near future.

**Comment by dathudeptrai** [comment]
@AyanmoI awsome !. The current TF XLA for text generation is one of the fastest engine out there. If we have flash attention, that is a big win :D.

**Comment by AyanmoI** [comment]
> @AyanmoI awsome !. The current TF XLA for LLM is one of the fastest engine out there. If we have flash attention, that is a big win :D.

That is good to know :)

**Pull request closed without merging**